runtime/lib/capi/ExecutionContext.hpp

```diff
@@ -40,38 +40,59 @@ extern "C" void __catalyst_inactive_callback(int64_t identifier, int64_t argc, i
 
 class MemoryManager final {
   private:
-    std::unordered_set<void *> _impl;
+    std::unordered_set<std::shared_ptr<void>> _impl;
     std::mutex mu; // To guard the memory manager
 
   public:
     explicit MemoryManager() { _impl.reserve(1024); };
 
-    ~MemoryManager()
+    std::shared_ptr<void> create(size_t size)
     {
-        // Lock the mutex to protect _impl free
-        std::lock_guard<std::mutex> lock(mu);
-        for (auto allocation : _impl) {
-            free(allocation);
-        }
+        std::shared_ptr<void> p(std::malloc(size), [](void *ptr) { free(ptr); });
+        this->insert(p);
+        return p;
     }
 
-    void insert(void *ptr)
+    std::shared_ptr<void> create_aligned(size_t alignment, size_t size)
+    {
+        std::shared_ptr<void> p(std::aligned_alloc(alignment, size), [](void *ptr) { free(ptr); });
+        this->insert(p);
+        return p;
+    }
+
+    void insert(std::shared_ptr<void> ptr)
     {
         // Lock the mutex to protect _impl update
         std::lock_guard<std::mutex> lock(mu);
         _impl.insert(ptr);
     }
+
     void erase(void *ptr)
     {
         // Lock the mutex to protect _impl update
         std::lock_guard<std::mutex> lock(mu);
-        _impl.erase(ptr);
+
+        std::shared_ptr<void> target;
+        for (std::shared_ptr<void> sharedP : _impl) {
+            if (sharedP.get() == ptr) {
+                target = sharedP;
+            }
+        }
+        _impl.erase(target);
```

Comment:
Doesn't the implementation of erase will call the destructor of target? And since target is a std::shared_ptr it will decrement its reference count? This goes against what we want erase to do, which is to erase the pointer of the set which needs to be deleted.

Reply:
Ah, so this class is kind of like the anti-smart_ptr, where we hold a "dead" pointer, waiting for it to potentially be recycled, and manually free the pointer once the holder itself goes out of scope?

I wasn't familiar with the context here and was just trying to fix clang tidy warnings, but now that I know I think this is one instance where we have to manually silence clang tidy. I reckon this can be closed then?

Reply:
> Ah, so this class is kind of like the anti-smart_ptr, where we hold a "dead" pointer, waiting for it to potentially be recycled, and manually free the pointer once the holder itself goes out of scope?

No. It is like a unique_ptr. But with a way to not free the memory. If we knew they were dead, we would delete them. Essentially:

1. Elements in the set may or may not be alive. When `mlir_memref_free` the pointer is removed from this set and freed. 
2. When the set is destroyed, everything in the set is freed. (Imagine it like every element in the set is reference counted by a single counter, and here we decrement that counter).
3. You can remove elements from the set without free-ing. This is useful to transfer the ownership of memory (and technically not what an arena allocator is like).

I call this an "arena-like" allocator, but it is not exactly an arena allocator.

Reply:
Just as a note, this design you are proposing may work if you return a shared_pointer but erase it from the set. It is then up to the caller to keep the pointer alive.